vShared plugin: volume mount failed due to cannot start service before timeout

Steps:
1. create a volume with vShared plugin
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create --driver=vshared --name=vol1
vol1
root@sc-rdops-vm02-dhcp-52-237:~# docker volume inspect vol1
[
    {
        "Driver": "vshared:latest",
        "Labels": {},
        "Mountpoint": "/mnt/vshared/vol1/",
        "Name": "vol1",
        "Options": {},
        "Scope": "global",
        "Status": {
            "Clients": null,
            "File server Port": 0,
            "Global Refcount": 0,
            "Service name": "",
            "Volume Status": "Ready"
        }
    }
]

```
2. run a container to attach this volume
```
root@sc-rdops-vm02-dhcp-52-237:~# docker run -it -v vol1:/vol busybox /bin/sh
Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
9e87eff13613: Pull complete 
Digest: sha256:2605a2c4875ce5eb27a9f7403263190cd1af31e48a2044d400320548356251c4
Status: Downloaded newer image for busybox:latest
docker: Error response from daemon: error while mounting volume '/mnt/vshared/vol1/': VolumeDriver.Mount: Failed to blocking wait for Mounted state. Error: Timeout reached; BlockingWait is not complete..

root@sc-rdops-vm02-dhcp-52-237:~# docker volume inspect vol1
[
    {
        "Driver": "vshared:latest",
        "Labels": {},
        "Mountpoint": "/mnt/vshared/vol1/",
        "Name": "vol1",
        "Options": {},
        "Scope": "global",
        "Status": {
            "Clients": null,
            "File server Port": 0,
            "Global Refcount": 0,
            "Service name": "",
            "Volume Status": "Error"
        }
    }
]

```

This is the vsphere_shared.log:
```
2017-08-16 15:37:34.151414591 -0700 PDT [INFO] VolumeDriver Get: vol1
2017-08-16 15:37:34.183235052 -0700 PDT [INFO] Volume not found: vol1
2017-08-16 15:37:34.184316907 -0700 PDT [ERROR] Failed to get volume meta-data error="No such volume" name=vol1 
2017-08-16 15:37:34.185197406 -0700 PDT [INFO] VolumeDriver Create: vol1
2017-08-16 15:37:34.185492167 -0700 PDT [INFO] Attempting to write initial metadata entry for vol1
2017-08-16 15:37:34.205793526 -0700 PDT [INFO] Watcher on global refcount returns event type=PUT 
2017-08-16 15:37:34.206888201 -0700 PDT [INFO] Attempting to create internal volume for vol1
2017-08-16 15:37:39.513496693 -0700 PDT [INFO] Attempting to update volume state to ready for volume: vol1
2017-08-16 15:37:39.594893479 -0700 PDT [INFO] Successfully created volume: vol1
2017-08-16 15:37:47.723957447 -0700 PDT [INFO] VolumeDriver Get: vol1
2017-08-16 15:38:01.259606622 -0700 PDT [INFO] VolumeDriver Get: vol1
2017-08-16 15:38:01.532861749 -0700 PDT [INFO] Mounting volume name=vol1 
2017-08-16 15:38:01.533706618 -0700 PDT [INFO] Directory doesn't exist, creating it path="/mnt/vshared/vol1/" 
2017-08-16 15:38:01.533827245 -0700 PDT [INFO] Before AtomicIncr
2017-08-16 15:38:02.599618816 -0700 PDT [INFO] Watcher on global refcount returns event type=PUT 
2017-08-16 15:38:03.600109932 -0700 PDT [INFO] Attempting to change volume state to Mounting
2017-08-16 15:38:10.304162321 -0700 PDT [INFO] Checking status of file server container...
2017-08-16 15:38:10.318087448 -0700 PDT [INFO] File server not running for volume vol1
2017-08-16 15:38:15.30415672 -0700 PDT [INFO] Checking status of file server container...
2017-08-16 15:38:15.308395133 -0700 PDT [INFO] File server not running for volume vol1
2017-08-16 15:38:20.304441888 -0700 PDT [INFO] Checking status of file server container...
2017-08-16 15:38:20.308937468 -0700 PDT [INFO] File server not running for volume vol1
2017-08-16 15:38:25.30413009 -0700 PDT [INFO] Checking status of file server container...
2017-08-16 15:38:25.307482481 -0700 PDT [INFO] File server not running for volume vol1
2017-08-16 15:38:30.304159117 -0700 PDT [INFO] Checking status of file server container...
2017-08-16 15:38:30.307000833 -0700 PDT [INFO] File server not running for volume vol1
2017-08-16 15:38:35.304201001 -0700 PDT [INFO] Checking status of file server container...
2017-08-16 15:38:35.30726231 -0700 PDT [INFO] File server not running for volume vol1
2017-08-16 15:38:35.307351514 -0700 PDT [WARNING] Timeout reached while waiting for file server container for volume vol1
2017-08-16 15:38:43.715999508 -0700 PDT [INFO] Watcher on global refcount returns event type=PUT 
2017-08-16 15:38:43.716619966 -0700 PDT [ERROR] name=vol1 error="Failed to blocking wait for Mounted state. Error: Timeout reached; BlockingWait is not complete." 
2017-08-16 15:38:43.716865852 -0700 PDT [ERROR] Failed to mount name=vol1 error="Failed to blocking wait for Mounted state. Error: Timeout reached; BlockingWait is not complete." 
```

docker service takes longer time than expected to start.
```
root@sc-rdops-vm02-dhcp-52-237:~# docker service ls
ID            NAME            MODE        REPLICAS  IMAGE
zlzxbqm0s728  vol1fileServer  replicated  0/1       dperson/samba:latest
root@sc-rdops-vm02-dhcp-52-237:~# docker service ps vol1fileServer 
ID            NAME                  IMAGE                 NODE                       DESIRED STATE  CURRENT STATE             ERROR                             PORTS
kbihay4nds7k  vol1fileServer.1      dperson/samba:latest  sc-rdops-vm02-dhcp-52-237  Running        Preparing 14 seconds ago                                    
kca4wdczcpp2   \_ vol1fileServer.1  dperson/samba:latest  sc-rdops-vm02-dhcp-52-237  Shutdown       Rejected 15 seconds ago   "No such image: dperson/samba@…"  
root@sc-rdops-vm02-dhcp-52-237:~# docker service ls
ID            NAME            MODE        REPLICAS  IMAGE
zlzxbqm0s728  vol1fileServer  replicated  1/1       dperson/samba:latest

```

--- a/client_plugin/drivers/shared/dockerops/dockerops.go
+++ b/client_plugin/drivers/shared/dockerops/dockerops.go
@@ -66,7 +66,7 @@ const (
 	// Time between successive checks for deleting a volume
 	checkSleepDuration = time.Second
 	// Timeout to mark Samba service launch as unsuccessful
-	sambaRequestTimeout = 30 * time.Second
+	sambaRequestTimeout = 150 * time.Second
 	// Prefix for internal volume names
 	internalVolumePrefix = "InternalVol"
 	// Error returned when no Samba service for that volume exists

--- a/client_plugin/drivers/shared/dockerops/dockerops.go
+++ b/client_plugin/drivers/shared/dockerops/dockerops.go
@@ -67,7 +67,7 @@ const (
 	// Time between successive checks for deleting a volume
 	checkSleepDuration = time.Second
 	// default Timeout to mark Samba service launch as unsuccessful
-	defaultServiceStartTimeOutInSecond = 30
+	defaultSvcStartTimeoutSec = 30
 	// Prefix for internal volume names
 	internalVolumePrefix = "InternalVol"
 	// Error returned when no Samba service for that volume exists
@@ -78,7 +78,7 @@ const (
 func GetServiceStartTimeout() time.Duration {
 	timeOutSec, err := strconv.Atoi(os.Getenv("VFILE_TIMEOUT_IN_SECOND"))
 	if err != nil {
-		timeOutSec = defaultServiceStartTimeOutInSecond
+		timeOutSec = defaultSvcStartTimeoutSec
 	}
 	log.WithFields(log.Fields{"value": timeOutSec}).Info("Service start timeout")
 	return time.Duration(timeOutSec) * time.Second

--- a/client_plugin/drivers/shared/dockerops/dockerops.go
+++ b/client_plugin/drivers/shared/dockerops/dockerops.go
@@ -67,41 +67,21 @@ const (
 	// Time between successive checks for deleting a volume
 	checkSleepDuration = time.Second
 	// default Timeout to mark Samba service launch as unsuccessful
-	defaultSambaRequestTimeOutInSecond = 30
+	defaultServiceStartTimeOutInSecond = 30
 	// Prefix for internal volume names
 	internalVolumePrefix = "InternalVol"
 	// Error returned when no Samba service for that volume exists
 	noSambaServiceError = "No file service exists"
 )
 
-var (
-	SambaRequestTimeout      time.Duration
-	isSambaRequestTimeoutSet bool = false
-)
-
-// return SambaRequest timeout value
-func GetSambaRequestTimeout() time.Duration {
-	if isSambaRequestTimeoutSet != false {
-		return SambaRequestTimeout
+// get service start timeout
+func GetServiceStartTimeout() time.Duration {
+	timeOutSec, err := strconv.Atoi(os.Getenv("VFILE_TIMEOUT_IN_SECOND"))
+	if err != nil {
+		timeOutSec = defaultServiceStartTimeOutInSecond
 	}
-	// read sambaRequest Timout value from enviroment variable
-	timeOutFromConfig := os.Getenv("VFILE_TIMEOUT_IN_SECOND")
-	val, err := strconv.Atoi(timeOutFromConfig)
-	if err == nil {
-		log.WithFields(log.Fields{
-			"value": val,
-		}).Info("Get sambaRequest timeout from config ")
-
-		SambaRequestTimeout = time.Duration(val) * time.Second
-	} else {
-		log.WithFields(log.Fields{
-			"value": defaultSambaRequestTimeOutInSecond,
-		}).Info("Use default value for sambaRequest timeout ")
-
-		SambaRequestTimeout = time.Duration(defaultSambaRequestTimeOutInSecond) * time.Second
-	}
-	isSambaRequestTimeoutSet = true
-	return SambaRequestTimeout
+	log.WithFields(log.Fields{"value": timeOutSec}).Info("Service start timeout")
+	return time.Duration(timeOutSec) * time.Second
 }
 
 // DockerOps is the interface for docker host related operations
@@ -305,7 +285,7 @@ func (d *DockerOps) StartSMBServer(volName string) (int, string, bool) {
 	// Wait till service container starts
 	ticker := time.NewTicker(checkDuration)
 	defer ticker.Stop()
-	timer := time.NewTimer(GetSambaRequestTimeout())
+	timer := time.NewTimer(GetServiceStartTimeout())
 	defer timer.Stop()
 	for {
 		select {
@@ -461,9 +441,9 @@ func (d *DockerOps) DeleteInternalVolume(volName string) {
 	internalVolname := internalVolumePrefix + volName
 	ticker := time.NewTicker(checkSleepDuration)
 	defer ticker.Stop()
-	// timeout set to sambaRequestTimeout because the internal volume maybe
+	// timeout set to the service start timeout because the internal volume maybe
 	// still in use due to stop of SMB server in progress
-	timer := time.NewTimer(sambaRequestTimeout)
+	timer := time.NewTimer(GetServiceStartTimeout())
 	defer timer.Stop()
 
 	for {
@@ -522,7 +502,7 @@ func (d *DockerOps) StopSMBServer(volName string) (int, string, bool) {
 	// Wait till service container stops
 	ticker := time.NewTicker(checkDuration)
 	defer ticker.Stop()
-	timer := time.NewTimer(GetSambaRequestTimeout())
+	timer := time.NewTimer(GetServiceStartTimeout())
 	defer timer.Stop()
 	for {
 		select {

--- a/client_plugin/drivers/shared/dockerops/dockerops.go
+++ b/client_plugin/drivers/shared/dockerops/dockerops.go
@@ -27,7 +27,11 @@ import (
 	"fmt"
 	"io"
 	"os"
+<<<<<<< HEAD
 	"strings"
+=======
+	"strconv"
+>>>>>>> Address comments from Mark to make the sambaRequest timeout value as a config option.
 	"time"
 
 	log "github.com/Sirupsen/logrus"
@@ -63,15 +67,50 @@ const (
 	// Time between successive checks for Samba service
 	// status to see if service container was launched
 	checkDuration = 5 * time.Second
+<<<<<<< HEAD
 	// Time between successive checks for deleting a volume
 	checkSleepDuration = time.Second
 	// Timeout to mark Samba service launch as unsuccessful
 	sambaRequestTimeout = 150 * time.Second
+=======
+	// default Timeout to mark Samba service launch as unsuccessful
+	defaultSambaRequestTimeOutInSecond = 30
+>>>>>>> Address comments from Mark to make the sambaRequest timeout value as a config option.
 	// Prefix for internal volume names
 	internalVolumePrefix = "InternalVol"
 	// Error returned when no Samba service for that volume exists
 	noSambaServiceError = "No file service exists"
 )
+
+var (
+	SambaRequestTimeout      time.Duration
+	isSambaRequestTimeoutSet bool = false
+)
+
+// return SambaRequest timeout value
+func GetSambaRequestTimeout() time.Duration {
+	if isSambaRequestTimeoutSet != false {
+		return SambaRequestTimeout
+	}
+	// read sambaRequest Timout value from enviroment variable
+	timeOutFromConfig := os.Getenv("VFILE_TIMEOUT_IN_SECOND")
+	val, err := strconv.Atoi(timeOutFromConfig)
+	if err == nil {
+		log.WithFields(log.Fields{
+			"value": val,
+		}).Info("Get sambaRequest timeout from config ")
+
+		SambaRequestTimeout = time.Duration(val) * time.Second
+	} else {
+		log.WithFields(log.Fields{
+			"value": defaultSambaRequestTimeOutInSecond,
+		}).Info("Use default value for sambaRequest timeout ")
+
+		SambaRequestTimeout = time.Duration(defaultSambaRequestTimeOutInSecond) * time.Second
+	}
+	isSambaRequestTimeoutSet = true
+	return SambaRequestTimeout
+}
 
 // DockerOps is the interface for docker host related operations
 type DockerOps struct {
@@ -274,7 +313,7 @@ func (d *DockerOps) StartSMBServer(volName string) (int, string, bool) {
 	// Wait till service container starts
 	ticker := time.NewTicker(checkDuration)
 	defer ticker.Stop()
-	timer := time.NewTimer(sambaRequestTimeout)
+	timer := time.NewTimer(GetSambaRequestTimeout())
 	defer timer.Stop()
 	for {
 		select {
@@ -491,7 +530,7 @@ func (d *DockerOps) StopSMBServer(volName string) (int, string, bool) {
 	// Wait till service container stops
 	ticker := time.NewTicker(checkDuration)
 	defer ticker.Stop()
-	timer := time.NewTimer(sambaRequestTimeout)
+	timer := time.NewTimer(GetSambaRequestTimeout())
 	defer timer.Stop()
 	for {
 		select {

--- a/client_plugin/drivers/shared/dockerops/dockerops.go
+++ b/client_plugin/drivers/shared/dockerops/dockerops.go
@@ -27,11 +27,8 @@ import (
 	"fmt"
 	"io"
 	"os"
-<<<<<<< HEAD
-	"strings"
-=======
 	"strconv"
->>>>>>> Address comments from Mark to make the sambaRequest timeout value as a config option.
+	"strings"
 	"time"
 
 	log "github.com/Sirupsen/logrus"
@@ -67,15 +64,10 @@ const (
 	// Time between successive checks for Samba service
 	// status to see if service container was launched
 	checkDuration = 5 * time.Second
-<<<<<<< HEAD
 	// Time between successive checks for deleting a volume
 	checkSleepDuration = time.Second
-	// Timeout to mark Samba service launch as unsuccessful
-	sambaRequestTimeout = 150 * time.Second
-=======
 	// default Timeout to mark Samba service launch as unsuccessful
 	defaultSambaRequestTimeOutInSecond = 30
->>>>>>> Address comments from Mark to make the sambaRequest timeout value as a config option.
 	// Prefix for internal volume names
 	internalVolumePrefix = "InternalVol"
 	// Error returned when no Samba service for that volume exists

--- a/client_plugin/drivers/shared/kvstore/etcdops/etcdops.go
+++ b/client_plugin/drivers/shared/kvstore/etcdops/etcdops.go
@@ -963,7 +963,7 @@ func (e *EtcdKVS) BlockingWaitAndGet(key string, value string, newKey string) (s
 	defer ticker.Stop()
 	// This call is used to block and wait for long
 	// running functions. Larger timeout is justified.
-	timer := time.NewTimer(dockerops.GetSambaRequestTimeout() + 2*requestTimeout)
+	timer := time.NewTimer(dockerops.GetServiceStartTimeout() + 2*requestTimeout)
 	defer timer.Stop()
 
 	for {

--- a/client_plugin/drivers/shared/kvstore/etcdops/etcdops.go
+++ b/client_plugin/drivers/shared/kvstore/etcdops/etcdops.go
@@ -963,7 +963,7 @@ func (e *EtcdKVS) BlockingWaitAndGet(key string, value string, newKey string) (s
 	defer ticker.Stop()
 	// This call is used to block and wait for long
 	// running functions. Larger timeout is justified.
-	timer := time.NewTimer(8 * requestTimeout)
+	timer := time.NewTimer(24 * requestTimeout)
 	defer timer.Stop()
 
 	for {

--- a/client_plugin/drivers/shared/kvstore/etcdops/etcdops.go
+++ b/client_plugin/drivers/shared/kvstore/etcdops/etcdops.go
@@ -963,7 +963,7 @@ func (e *EtcdKVS) BlockingWaitAndGet(key string, value string, newKey string) (s
 	defer ticker.Stop()
 	// This call is used to block and wait for long
 	// running functions. Larger timeout is justified.
-	timer := time.NewTimer(24 * requestTimeout)
+	timer := time.NewTimer(dockerops.GetSambaRequestTimeout() + 2*requestTimeout)
 	defer timer.Stop()
 
 	for {

--- a/docs/user-guide/vfile-plugin.md
+++ b/docs/user-guide/vfile-plugin.md
@@ -3,7 +3,7 @@ title: vFile volume plugin for Docker
 
 ---
 ## Overview
-Depending on the underlying block storage device system, it might not be possible to access the same
+Depending on the underlying block storage system, it might not be possible to access the same
 persistent volume across different hosts/nodes simultanously.
 For example, currently users cannot mount the same persistent volume which is created through
 vSphere Docker Volume Service (vDVS) on containers running on two different hosts at the same time.
@@ -138,9 +138,9 @@ When you see somthing like the following in the log
 Please make sure the volume you used is a valid volume name. A valid volume name consists of ```[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]```.
 
 ### I got " VolumeDriver.Mount: Failed to blocking wait for Mounted state. Error: Timeout reached; BlockingWait is not complete.." when mounting a volume.
-We see this issue only on platforms which the availble disk space are low. On those platforms, docker swarm service takes longer time to up and running. We recommend to run this driver on platforms with enough available disk space. You can also try to increase the service start timeout value by specifing ```VFILE_TIMEOUT_IN_SECOND``` value when installing the plugin:
+We see this issue only on platform where the space is low. When available disk space are low, Docker Swarm service may take longer to start a service. Generally it's better free up some disk space. You can also try to increase the service start timeout value, controlled by ```VFILE_TIMEOUT_IN_SECOND``` env variable:
 ```
 docker plugin install --grant-all-permissions --alias vfile cnastorage/vfile:latest VFILE_TIMEOUT_IN_SECOND=90
 ```
-The above command increases the service start timeout value to 90 seconds and the default value is 30 seconds.
+This will increase timeout to 90 sec, from default of 30 sec.
 

--- a/docs/user-guide/vfile-plugin.md
+++ b/docs/user-guide/vfile-plugin.md
@@ -144,3 +144,4 @@ docker plugin install --grant-all-permissions --alias vfile cnastorage/vfile:lat
 ```
 This will increase timeout to 90 sec, from default of 30 sec.
 
+

--- a/docs/user-guide/vfile-plugin.md
+++ b/docs/user-guide/vfile-plugin.md
@@ -144,4 +144,3 @@ docker plugin install --grant-all-permissions --alias vfile cnastorage/vfile:lat
 ```
 This will increase timeout to 90 sec, from default of 30 sec.
 
-

--- a/docs/user-guide/vfile-plugin.md
+++ b/docs/user-guide/vfile-plugin.md
@@ -9,7 +9,11 @@ For example, currently users cannot mount the same persistent volume which is cr
 vSphere Docker Volume Service (vDVS) on containers running on two different hosts at the same time.
 
 This can be solved through distributed file systems, such as NFS, Ceph, Gluster, etc.
+<<<<<<< HEAD
 However, setting up and maintaining enterprise storage offerings for Cloud Native usecases is not a trivial work.
+=======
+However, setting up and maintaining those distributed file systems for docker persistent data usage is not a trivial work.
+>>>>>>> Update user guide.
 Furthermore, users can face more challenges in order to achieve high availability, scalability, and load balancing.
 
 __vFile volume plugin for Docker__ provides simultanous persistent volume access between hosts in the
@@ -132,3 +136,11 @@ When you see somthing like the following in the log
 2017-08-24 11:57:16.436786459 -0700 PDT [WARNING] Failed to create file server for volume space vol7. Reason: Error response from daemon: {"message":"rpc error: code = 3 desc = name must be valid as a DNS name component"}
 ```
 Please make sure the volume you used is a valid volume name. A valid volume name consists of ```[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]```.
+
+### I got " VolumeDriver.Mount: Failed to blocking wait for Mounted state. Error: Timeout reached; BlockingWait is not complete.." when mounting a volume.
+We see this issue only on platforms which the availble disk space are low. On those platforms, docker swarm service takes longer time to up and running. We recommend to run this driver on platforms with enough available disk space. You can also try to increase the service start timeout value by specifing ```VFILE_TIMEOUT_IN_SECOND``` value when installing the plugin:
+```
+docker plugin install --grant-all-permissions --alias vfile cnastorage/vfile:latest VFILE_TIMEOUT_IN_SECOND=90
+```
+The above command increases the service start timeout value to 90 seconds and the default value is 30 seconds.
+

--- a/plugin_dockerbuild/config.json-template
+++ b/plugin_dockerbuild/config.json-template
@@ -58,6 +58,12 @@
 		"description": "Client protocol version used in test",
 		"value": "",
 		"Settable": [ "value"]
+	},
+	{
+		"name": "VFILE_TIMEOUT_IN_SECOND",
+		"description": "Timeout value in second used by vFILE plugin",
+		"value": "",
+		"Settable": [ "value"]
 	}
 	]
 }

